### PR TITLE
Add RDT Memory Bandwidth Monitoring (MBM) and Cache Monitoring Technology (CMT) statistics.

### PR DESCRIPTION
**Pull request by Creatone**
This pull request introduces the reading of Intel RDT Memory Bandwidth Monitoring statistics.

### CODE REVIEW

File: libcontainer/intelrdt/intelrdt.go [hunk 1]
```diff
@@ -298,20 +312,26 @@ func isIntelRdtMounted() bool {
 	return true
 }
 
-func parseCpuInfoFile(path string) (bool, bool, error) {
-	isCatFlagSet := false
-	isMbaFlagSet := false
+type cpuInfoFlags struct {
+	isCatFlagSet      bool
```
Comment (kolyshkin) [inline libcontainer/intelrdt/intelrdt.go:316]: You can probably drop the word `Flag`, i.e. `isCatSet`, `isMbaSet` etc.

Come to think of it, you can also drop `is` and `Set`, too.

**Comment by kolyshkin** [comment]
@Creatone can you please avoid the merge commit? You can do so by using `git pull --rebase` instead of `git rebase`.

File: libcontainer/intelrdt/intelrdt.go [hunk 2]
```diff
@@ -298,20 +312,26 @@ func isIntelRdtMounted() bool {
...
+	isMbaFlagSet      bool
+	isMbmTotalFlagSet bool
+	isMbmLocalFlagSet bool
+}
+
+func parseCpuInfoFile(path string) (cpuInfoFlags, error) {
+	infoFlags := cpuInfoFlags{false, false, false, false}
```
Comment (kolyshkin) [inline libcontainer/intelrdt/intelrdt.go:323]: You don't have to explicitly set `bool` fields to `false` -- it is a default value.

File: libcontainer/intelrdt/intelrdt.go [hunk 3]
```diff
@@ -298,20 +312,26 @@ func isIntelRdtMounted() bool {
 	return true
 }
 
-func parseCpuInfoFile(path string) (bool, bool, error) {
```
Comment (kolyshkin) [inline libcontainer/intelrdt/intelrdt.go:301]: For the sake of easier review and better git history, can you please split this change into a separate commit?

In other words
 1. "intelrdt: change parseCpuInfoFile to return struct"
 2. "intelrdt: add mbm stats"

**Comment by Creatone** [comment]
@kolyshkin

File: libcontainer/intelrdt/mbm.go [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+		switch feature := scanner.Text(); feature {
+
+		case "mbm_total_bytes":
+			monFeatures.mbmTotalBytes = true
+		case "mbm_local_bytes":
+			monFeatures.mbmLocalBytes = true
+		case "llc_occupancy":
+			monFeatures.llcOccupancy = true
+		default:
+			logrus.Warn(fmt.Sprintf("Unsupported RDT Memory Bandwidth Monitoring (MBM) feature: %v", feature))
```
Comment (kolyshkin) [inline libcontainer/intelrdt/mbm.go:58]: 1. Please use `logrus.Warnf()`
2. `feature` is a string, so `%s` is preferred.

File: libcontainer/intelrdt/mbm.go [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+		}
+	}
+
+	if err := scanner.Err(); err != nil {
+		return monFeatures, err
+	}
+
+	return monFeatures, nil
```
Comment (kolyshkin) [inline libcontainer/intelrdt/mbm.go:66]: You can replace these 5 lines with just 
`return monFeatures, scanner.Err()`

File: libcontainer/intelrdt/mbm.go [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+}
+
+func getMBMStats(containerPath string) (*[]MBMNumaNodeStats, error) {
+	mbmStats := []MBMNumaNodeStats{}
+
+	numaPaths, err := filepath.Glob(filepath.Join(containerPath, "mon_data", "*"))
```
Comment (kolyshkin) [inline libcontainer/intelrdt/mbm.go:72]: Hmm, maybe use `ioutil.ReadDir(filepath.Join(containerPath, "mon_data")` instead? Not a strong preference though, I'm fine with either way.

File: libcontainer/intelrdt/mbm.go [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+	llcOccupancy  bool
+}
+
+// Check if Intel RDT/MBM is enabled
+func IsMbmEnabled() bool {
+	return isMbmEnabled
+}
+
+func getMonFeatures(intelRdtRoot string) (monFeatures, error) {
+	file, err := os.Open(filepath.Join(intelRdtRoot, "info", "L3_MON", "mon_features"))
```
Comment (kolyshkin) [inline libcontainer/intelrdt/mbm.go:34]: Actually, since we're on Linux, you can use `filepath.Join(intelRdtRoot, "info/L3_MON/mon_features")` or even `intelRdtRoot + "/info/L3_MON/mon_features"`.

File: libcontainer/intelrdt/mbm_test.go [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+		if err != nil {
+			t.Errorf("Error while parsing mon features err = %v", err)
+		}
+
+		expectedMonFeatures := monFeatures{false, false, false}
+
+		if parsedMonFeatures != expectedMonFeatures {
+			t.Errorf("Expected no features available but there is any!")
+		}
+
```
Comment (kolyshkin) [inline libcontainer/intelrdt/mbm_test.go:42]: no need for empty line here

File: libcontainer/intelrdt/mbm_test.go [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+	for _, numa := range NUMANodes {
+		numaPath := filepath.Join(monDataPath, numa)
+		err = os.MkdirAll(numaPath, os.ModePerm)
+		if err != nil {
+			return "", err
+		}
+
+		for fileName, value := range mocks {
+			err = prepareMockedFile(filepath.Join(numaPath, fileName), strconv.FormatUint(value, 10))
```
Comment (kolyshkin) [inline libcontainer/intelrdt/mbm_test.go:81]: you can use `ioutils.WriteFile()`

File: libcontainer/intelrdt/mbm_test.go [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+	mocksNUMANodesToCreate := []string{"mon_l3_00", "mon_l3_01"}
+
+	mocksFilesToCreate := map[string]uint64{
+		"mbm_total_bytes": 9123911,
+		"mbm_local_bytes": 2361361,
+		"llc_occupancy":   123013,
+	}
+
+	mockedMBM, err := mockMBM(mocksNUMANodesToCreate, mocksFilesToCreate)
+	defer cleanup(mockedMBM, t)
```
Comment (kolyshkin) [inline libcontainer/intelrdt/mbm_test.go:109]: no need to have a cleanup() function defined if you're only using it once.

instead, do
```go
defer func() {
        _ = os.RemoveAll(mockedMBM)
}()
```

(and there's no need to check for error since you can't do anything about it and this should not result in a test failure.

File: libcontainer/intelrdt/mbm_test.go [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+		enabledMonFeatures.mbmLocalBytes = true
+		enabledMonFeatures.llcOccupancy = true
+
+		stats, err := getMBMStats(mockedMBM)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		if len(*stats) != len(mocksNUMANodesToCreate) {
+			t.Fatal(
```
Comment (kolyshkin) [inline libcontainer/intelrdt/mbm_test.go:125]: please use `t.Fatalf()`

Comment (kolyshkin) [reply libcontainer/intelrdt/mbm_test.go:125]: in the places below, too

File: libcontainer/intelrdt/mbm.go [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+		switch feature := scanner.Text(); feature {
+
+		case "mbm_total_bytes":
+			monFeatures.mbmTotalBytes = true
+		case "mbm_local_bytes":
+			monFeatures.mbmLocalBytes = true
+		case "llc_occupancy":
+			monFeatures.llcOccupancy = true
+		default:
+			logrus.Warnf(fmt.Sprintf("Unsupported RDT Memory Bandwidth Monitoring (MBM) feature: %s", feature))
```
Comment (kolyshkin) [inline libcontainer/intelrdt/mbm.go:59]: :see_no_evil: please remove fmt.Sprintf

**Comment by Creatone** [comment]
@AkihiroSuda Could you review this?

File: types/events.go [hunk 13]
```diff
@@ -113,6 +115,12 @@ type IntelRdt struct {
 
 	// The memory bandwidth schema in 'container_id' group
 	MemBwSchema string `json:"mem_bw_schema,omitempty"`
+
+	// The memory bandwidth monitoring statistics from NUMA nodes in 'container_id' group
+	MBMStats *[]intelrdt.MBMNumaNodeStats `json:"mbm_statistics,omitempty"`
```
Comment (AkihiroSuda) [inline types/events.go:120]: maybe: `json:"mbm_stats,omitempty"`

**Comment by AkihiroSuda** [comment]
Looks good but a single nit

File: types/events.go (hunk 13 above)
Comment (Creatone) [reply types/events.go]: good idea

**Comment by AkihiroSuda** [comment]
LGTM

**Comment by Creatone** [comment]
@cyphar Could you review this?

**Comment by Creatone** [comment]
@kolyshkin

File: libcontainer/intelrdt/stats.go [hunk 14]
```diff
@@ -33,6 +46,12 @@ type Stats struct {
 
 	// The memory bandwidth schema in 'container_id' group
 	MemBwSchema string `json:"mem_bw_schema,omitempty"`
+
+	// The memory bandwidth monitoring statistics from NUMA nodes in 'container_id' group
+	MBMStats *[]MBMNumaNodeStats `json:"mbm_statistics,omitempty"`
```
Comment (AkihiroSuda) [inline libcontainer/intelrdt/stats.go:51]: statistics -> stats?

**Comment by AkihiroSuda** [comment]
LGTM

**Comment by Creatone** [comment]
@kolyshkin Is it possible to merge this?

**Comment by AkihiroSuda** [comment]
@kolyshkin PTAL

**Comment by AkihiroSuda** [comment]
@kolyshkin LGTY?

File: types/events.go (hunk 13 above)
Comment (kolyshkin) [reply types/events.go]: @Creatone ^^^

**Comment by kolyshkin** [comment]
One small nit in json name (`stats` vs `statistics`) but I'm fine either way

LGTM

File: types/events.go (hunk 13 above)
Comment (Creatone) [reply types/events.go]: @kolyshkin this comment is outdated. String was already changed long time ago.
